Common/include/basic_types/ad_structure.hpp

```diff
@@ -676,6 +719,42 @@ FORCEINLINE void ResumePreaccumulation(bool wasActive) {
   SU2_OMP_SAFE_GLOBAL_ACCESS(PreaccEnabled = true;)
 }
 
+struct ErrorReport {
+  unsigned long ErrorCounter;
+  std::ofstream* out;
+};
+
+FORCEINLINE void ResetErrorCounter(struct ErrorReport* report) { report->ErrorCounter = 0; }
+
+FORCEINLINE void SetDebugReportFile(struct ErrorReport* report, std::ofstream* output_file) {
+  report->out = output_file;
+}
+
+FORCEINLINE unsigned long* GetErrorCount(struct ErrorReport* report) { return &(report->ErrorCounter); }
+
+#ifdef CODI_TAG_TAPE
+
+FORCEINLINE void SetTag(int tag) { AD::getTape().setCurTag(tag); }
+FORCEINLINE void ClearTagOnVariable(su2double& v) { AD::getTape().clearTagOnVariable(v); }
+
+static void tagErrorCallback(int const& correctTag, int const& wrongTag, void* userData) {
+  ErrorReport* report = (ErrorReport*)userData;
```

Comment:
or reinterpret _cast if really necessary
```suggestion
  auto* report = static_cast<ErrorReport*>(userData);
```